src/services/block_service.cpp

```diff
@@ -79,10 +80,38 @@ void block_service::work()
     if (!started(bind(xpub, xsub)))
         return;
 
-    // TODO: tap in to failure conditions, such as high water.
-    // BUGBUG: stop is insufficient to stop the worker, because of relay().
-    // Relay messages between subscriber and publisher (blocks on context).
-    relay(xpub, xsub);
+    if (settings_.block_websocket_service_enabled)
+    {
+        // Bind to the websocket service endpoint.
+        const auto endpoint = settings_.block_websocket_endpoint(secure_);
+        if (!start_web(endpoint, secure_, "block"))
+            return;
+
+        zmq::poller poller;
+        poller.add(xpub);
+        poller.add(xsub);
+
+        while (!poller.terminated() && !stopped())
+        {
+            static constexpr size_t wait_milliseconds = 10;
+            const auto signaled = poller.wait(wait_milliseconds);
+
+            if (signaled.contains(xpub.id()))
+                forward(xpub, xsub);
+
+            if (signaled.contains(xsub.id()))
+                forward(xsub, xpub);
+
+            poll(wait_milliseconds);
+        }
+    }
+    else
```

Comment:
In the various endpoints I do not understand why the zeromq relay is modified. Binding the websocket endpoint should be entirely independent.

Reply:
The relay call blocks and is great for zmq alone.  But we now have an entirely different event loop that needs to be integrated somehow.  To avoid adding a thread that polls the websocket endpoint in a dedicated manner, the poll call is integrated (so you'll see for tx/block we do both the zmq poll, followed by the websocket poll; heartbeat doesn't need zmq poll, so it's replaced with the websocket poll).  A bit messy, I agree.

Reply:
In the above loop I don’t see any websocket code??

Reply:
Or is that the poll() method? (On my phone.)

Reply:
In any case it’s very inelegant. The zeromq wait loop is notification with timeout. The addition of a websocket poll (with or without notification) will slow the underlying service response by the full amount of its timeout.

Reply:
The only resolution I see is the creation of similar service instances to monitor the websocket endpoints on distinct logical threads. This will leave the existing endpoints unmodified and decouple their waits.

Reply:
Yes, it's the poll method.  And I don't mind the separate thread monitoring (an earlier revision did this, but does require thread safety since the zmq responses will still need to be communicated to the web code (e.g. publish_transaction will need to place something on a queue that the web thread can get, decode, and broadcast)).  The most inelegant is still query though.  We need a way to get things in and out, and unless we make a distinct zmq query interface with dedicated query workers for web requests, we can't easily separate these.  Thoughts on that?

Reply:
Unless you mean we double up on the existing services and one is derived from web_sockets and the other is not?  Then the event loops can be full distinct.  I'll have to think about that, but I think that could work.

Reply:
Zeromq is designed with this purpose in mind, it is a concurrency library as much as networking. The websocket endpoints are clients to the service endpoints. They operate on their own thread and messages are related by zeromq.

Reply:
It would also be possible to relay directly from the websocket endpoints to the workers, but that would prove to be a more complex topology.